include/zephyr/location.h

```diff
@@ -0,0 +1,402 @@
+/**
+ * @file location.h
+ *
+ * @brief Public location API.
+ *
+ */
+
+/*
+ * Copyright (c) 2022, Bjarki Arge Andreasen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_LOCATION_H_
+#define ZEPHYR_INCLUDE_LOCATION_H_
+
+/**
+ * @brief Location Interface
+ * @defgroup location_interface Location Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <zephyr/device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Location data
+ *
+ * The latitude and longitude are in the unit micro degree. 1 micro
+ * degree corresponds to approx. 11cm along the earths surface.
+ *
+ * The accuracy describes the radius within which the real
+ * position is with a 68.2% certainty, in millimeters.
+ */
+struct location_position {
+	/** Latitudal position in microdegrees (0 to +-180000000) */
+	int32_t latitude;
+	/** Longitudal position in microdegrees (0 to +-180000000) */
+	int32_t longitude;
+	/** The uptime when the position was determined */
+	int64_t uptime;
```

Comment:
If you're using an `int64_t` to store timestamps, you might as well store the complete tick resolution.
`k_uptime_ticks()` instead of `k_uptime_get()`. Rename the variable to `uptime_ticks`?

Reply:
According to https://docs.zephyrproject.org/latest/kernel/services/timing/clocks.html?highlight=uptime#uptime it is recommended to use k_uptime_get() if the system does not have precise timing requirements, and if there is the possibility that the app will run longer than 2³² milliseconds.

I will argue that the uptime is not a precise timing instrument, as the timing accuracy from sources like GNSS modems is  generally in the ms scale (The NMEA GPZDA message stores time in hhmmss.ss for example). The same is true for getting a time over the internet (because of time from server to unit etc.). This is without accounting for the time it takes to transmit the data from the modem over UART for example. At baud 9600, it takes >30ms to send the GPZDA message for example.

I prefer using milliseconds since it is a real time unit, which makes it more human readable and faster since it requires no conversion from ticks, and since the utility of the uptime is not diminished due to the precision of the uptime being less than 1ms.

Reply:
I disagree on some of these points.

> which is the reason not to store it as uint32_t using k_uptime_get32()

I am not disagreeing with the decision to use a int64_t (although if it is changed to ticks as I suggest it should be `k_ticks_t`)

> At baud 9600, it takes >30ms to send the GPZDA message for example.

There is nothing to say that the timestamp needs to come from the UART port (at a horrifically slow baudrate). PPS is typically exposed from these modems, which could be timestamped to orders of magnitude more precision that the 32kHz clock rate. Further limiting it to a 1kHz clock rate is IMO silly.

> (The NMEA GPZDA message stores time in hhmmss.ss for example)

UBlox GNSS modems report timestamp accuracies on the order of nano-seconds.

> I prefer using milliseconds since it is a real time unit, which makes it more human readable and faster since it requires no conversion from ticks,

It is not faster than storing it in ticks, `k_uptime_get` is internally doing a conversion from `ticks` to `ms`, which is slower than just storing the tick counter directly. If you want a human readable uptime to print, use `k_ticks_to_ms_floor32` in your `printk` function. At least in my case, applications spend 99.999% of their time not connected to any form or serial output. Why spend time converting to worse resolution that is never going to be seen?

The higher resolution values here are useful for the `timeutil_sync` API.

Reply:
If you believe that it is possible to achieve that high timing resolution, then yes, it should be uptime_ticks. I don't believe it is possible, atleast yet, for reasons stated before. The PPS is accurate to the UTC timestamp, but not necessarily to the time the location data was created, which the NMEA message for example does indicate, at a lower resolution.

I will update it to use k_ticks_t, at the cost of having a need to use the k_ticks_to_ms_floor32 in the application space, which is what i really would love to be without if possible :)